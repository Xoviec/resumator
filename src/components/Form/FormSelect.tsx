import { VoidFunctionComponent } from "react";
import { Controller, useFormContext } from "react-hook-form";
import { OutlinedInput, Select, SelectProps } from "@mui/material";

type FormSelectProps = SelectProps & {
  name: string;
};

export const FormSelect: VoidFunctionComponent<FormSelectProps> = ({
  name,
  defaultValue,
  ...props
}) => {
  const { control } = useFormContext();

  return (
    <Controller
      name={name}
      control={control}
      defaultValue={defaultValue}
      render={({ field }) => (
        <Select
          fullWidth
          input={<OutlinedInput role="input" margin="dense" />}
          defaultValue={defaultValue}
          {...field}
          {...props}
        />
      )}
    />
  );
};
