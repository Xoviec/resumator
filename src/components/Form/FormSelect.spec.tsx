import { render, fireEvent } from "@testing-library/react";
import { FunctionComponent, useEffect } from "react";
import { FormProvider, useForm } from "react-hook-form";
import { FormSelect } from "./FormSelect";

jest.mock("@mui/material", () => ({
  ...jest.requireActual("@mui/material"),
  Select: jest
    .requireActual("react")
    .forwardRef(({ value, onChange }: any, ref: any) => (
      <input
        ref={ref}
        value={value || ""}
        onChange={onChange}
        data-testid="select-field"
      />
    )),
}));

describe("FormSelect", () => {
  it("should render correctly", () => {
    const FormProviderWrapper: FunctionComponent = ({ children }) => {
      const form = useForm();

      return <FormProvider {...form}>{children}</FormProvider>;
    };

    const { asFragment } = render(
      <FormProviderWrapper>
        <FormSelect name="test-form-name" value="test" />
      </FormProviderWrapper>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should update the value when selected", () => {
    const fieldName = "test-form-name";

    const FormProviderWrapper: FunctionComponent<any> = ({ children, onChange }) => {
      const form = useForm();
      const fieldValue = form.watch(fieldName);

      useEffect(() => {
        onChange(fieldValue);
      }, [fieldValue, onChange]);

      return <FormProvider {...form}>{children}</FormProvider>;
    };

    const mockChange = jest.fn();

    const { getByTestId } = render(
      <FormProviderWrapper onChange={mockChange}>
        <FormSelect name={fieldName} label="test-form-label" />
      </FormProviderWrapper>
    );

    fireEvent.change(getByTestId("select-field"), {
      target: { value: "test-input-value" },
    });

    expect(mockChange).toHaveBeenCalledWith("test-input-value");
  });
});
