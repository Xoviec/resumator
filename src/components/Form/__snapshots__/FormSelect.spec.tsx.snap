// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSelect should render correctly 1`] = `
<DocumentFragment>
  <input
    data-testid="select-field"
    value="test"
  />
</DocumentFragment>
`;
